streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -1032,13 +1039,25 @@ public void unregister(final Collection<TopicPartition> revokedChangelogs) {
                     // if the changelog is not in RESTORING, it means
                     // the corresponding onRestoreStart was not called; in this case
                     // we should not call onRestoreSuspended either
-                    if (changelogMetadata.stateManager.taskType() == Task.TaskType.ACTIVE &&
-                        changelogMetadata.state().equals(ChangelogState.RESTORING)) {
-                        try {
-                            final String storeName = changelogMetadata.storeMetadata.store().name();
-                            stateRestoreListener.onRestoreSuspended(partition, storeName, changelogMetadata.totalRestored);
-                        } catch (final Exception e) {
-                            throw new StreamsException("State restore listener failed on restore paused", e);
+                    if (changelogMetadata.state().equals(ChangelogState.RESTORING)) {
+                        final String storeName = changelogMetadata.storeMetadata.store().name();
+                        if (changelogMetadata.stateManager.taskType() == Task.TaskType.ACTIVE) {
+                            try {
+                                stateRestoreListener.onRestoreSuspended(partition, storeName, changelogMetadata.totalRestored);
+                            } catch (final Exception e) {
+                                throw new StreamsException("State restore listener failed on restore paused", e);
+                            }
+                        } else if (changelogMetadata.stateManager.taskType() == TaskType.STANDBY) {
+                            final StateStoreMetadata storeMetadata = changelogMetadata.stateManager.storeMetadata(partition);
+                            // endOffset and storeOffset may be unknown at this point
+                            final long storeOffset = storeMetadata.offset() != null ? storeMetadata.offset() : -1;
+                            final long endOffset = storeMetadata.endOffset() != null ? storeMetadata.endOffset() : -1;
+                            StandbyUpdateListener.SuspendReason suspendReason = StandbyUpdateListener.SuspendReason.MIGRATED;
+                            // Unregistering running standby tasks means the task has been promoted to active.
+                            if (changelogMetadata.stateManager.taskState() == Task.State.RUNNING) {
+                                suspendReason = StandbyUpdateListener.SuspendReason.PROMOTED;
+                            }
```

Comment:
nit: we always prefer using final variables to make the logic as clear as possible
```suggestion
                            // Unregistering running standby tasks means the task has been promoted to active.
                            final StandbyUpdateListener.SuspendReason suspendReason = 
                                changelogMetadata.stateManager.taskState() == Task.State.RUNNING 
                                    ? StandbyUpdateListener.SuspendReason.PROMOTED
                                    : StandbyUpdateListener.SuspendReason.MIGRATED;                        
```